samples/gltf_baker.cpp

```diff
@@ -53,98 +55,126 @@
+static const char* INI_FILENAME = "gltf_baker.ini";
+static const char* UV_FILENAME = "gltf_baker_tmp_uv.png";
+static const char* AO_FILENAME = "gltf_baker_tmp_ao.png";
+static constexpr int PATH_SIZE = 256;
 
-struct App {
-    Engine* engine;
-    Camera* camera;
-    SimpleViewer* viewer;
+struct BakerApp;
+
+using BakerAppTask = std::function<void(BakerApp*)>;
+
+struct BakerApp {
     Config config;
-    AssetLoader* loader;
-    FilamentAsset* asset = nullptr;
-    NameComponentManager* names;
-    MaterialProvider* materials;
-    bool actualSize = false;
-    AppState state = EMPTY;
+    Engine* engine = nullptr;
+    Camera* camera = nullptr;
+    SimpleViewer* viewer = nullptr;
+    NameComponentManager* names = nullptr;
+    MaterialProvider* materials = nullptr;
+    AssetLoader* loader = nullptr;
+    gltfio::AssetPipeline* pipeline = nullptr;
+    bool viewerActualSize = false;
     utils::Path filename;
+    bool hasTestRender = false;
+    bool isWorking = false;
+    std::string statusText;
+    std::string messageBoxText;
+    bool requestViewerUpdate = false;
+    Visualization visualization = Visualization::MESH_CURRENT;
+
+    // Bundle of Filament entities (renderables, textures, etc.) for the currently displayed mesh.
+    FilamentAsset* viewerAsset = nullptr;
+
+    // Available glTF scenes suitable for display, depending on "visualization".
+    gltfio::AssetPipeline::AssetHandle currentAsset = nullptr;
+    gltfio::AssetPipeline::AssetHandle modifiedAsset = nullptr;
+    gltfio::AssetPipeline::AssetHandle previewAoAsset = nullptr;
+    gltfio::AssetPipeline::AssetHandle previewUvAsset = nullptr;
+
+    // Available 2D images suitable for display, depending on "visualization".
     image::LinearImage ambientOcclusion;
     image::LinearImage bentNormals;
     image::LinearImage meshNormals;
     image::LinearImage meshPositions;
-    bool showOverlay = false;
-    bool enablePrepScale = true;
-    View* overlayView = nullptr;
-    Scene* overlayScene = nullptr;
-    VertexBuffer* overlayVb = nullptr;
-    IndexBuffer* overlayIb = nullptr;
-    Texture* overlayTexture = nullptr;
-    MaterialInstance* overlayMaterial = nullptr;
-    utils::Entity overlayEntity;
-    AppState pushedState;
-    gltfio::AssetPipeline* pipeline;
-    uint32_t bakeResolution = 1024;
-    ExportOption exportOption = PRESERVE_MATERIALS;
-
-    // Secondary threads might write to the following fields.
-    std::shared_ptr<std::string> statusText;
-    std::shared_ptr<std::string> messageBoxText;
-    std::atomic<bool> requestOverlayUpdate;
-    std::atomic<bool> requestStatePop;
-};
-
-struct OverlayVertex {
-    filament::math::float2 position;
-    filament::math::float2 uv;
-};
 
-static OverlayVertex OVERLAY_VERTICES[4] = {
-    {{0, 0}, {0, 0}},
-    {{ 1000, 0}, {1, 0}},
-    {{0,  1000}, {0, 1}},
-    {{ 1000,  1000}, {1, 1}},
+    // AssetPipeline callballs are triggered from outside the UI thread. To keep things simple, we
```

Comment:
"callbacks"

Reply:
Ben where would I be without you?

Reply:
I kind of want to find a reason to use the word callballs now :)